src/Client/ConnectionEstablisher.cpp

```diff
@@ -0,0 +1,236 @@
+    IConnectionPool * pool_,
+    const ConnectionTimeouts * timeouts_,
+    const Settings * settings_,
+    const QualifiedTableName * table_to_check_)
+    : pool(pool_), timeouts(timeouts_), settings(settings_), table_to_check(table_to_check_),
+      stage(Stage::INITIAL), log(&Poco::Logger::get("ConnectionEstablisher"))
+{
+#if defined(OS_LINUX)
+    epoll.add(receive_timeout.getDescriptor());
+#endif
+}
+
+void ConnectionEstablisher::Routine::ReadCallback::operator()(int fd, const Poco::Timespan & timeout, const std::string &)
+{
+#if defined(OS_LINUX)
+    if (connection_establisher.socket_fd != fd)
+    {
+        if (connection_establisher.socket_fd != -1)
+            connection_establisher.epoll.remove(connection_establisher.socket_fd);
+
+        connection_establisher.epoll.add(fd);
+        connection_establisher.socket_fd = fd;
+    }
+
+    connection_establisher.receive_timeout.setRelative(timeout);
+    fiber = std::move(fiber).resume();
+    connection_establisher.receive_timeout.reset();
+#else
+    (void) fd;
+    (void) timeout;
+#endif
+}
+
+Fiber ConnectionEstablisher::Routine::operator()(Fiber && sink)
+{
+    try
+    {
+        connection_establisher.establishConnection(ReadCallback{connection_establisher, sink});
+    }
+    catch (const boost::context::detail::forced_unwind &)
+    {
+        /// This exception is thrown by fiber implementation in case if fiber is being deleted but hasn't exited
+        /// It should not be caught or it will segfault.
+        /// Other exceptions must be caught
+        throw;
+    }
+    catch (...)
+    {
+        connection_establisher.exception = std::current_exception();
+    }
+
+    return std::move(sink);
+}
+
+void ConnectionEstablisher::resume()
+{
+    if (!fiber_created)
+    {
+        reset();
+        fiber = boost::context::fiber(std::allocator_arg_t(), fiber_stack, Routine{*this});
+        fiber_created = true;
+        resumeFiber();
+        return;
+    }
+
+#if defined(OS_LINUX)
+    bool is_socket_ready = false;
+    bool is_receive_timeout_alarmed = false;
+
+    epoll_event events[2];
+    events[0].data.fd = events[1].data.fd = -1;
+    size_t ready_count = epoll.getManyReady(2, events, true);
+    for (size_t i = 0; i != ready_count; ++i)
+    {
+        if (events[i].data.fd == socket_fd)
+            is_socket_ready = true;
+        if (events[i].data.fd == receive_timeout.getDescriptor())
+            is_receive_timeout_alarmed = true;
+    }
+
+    if (is_receive_timeout_alarmed && !is_socket_ready)
+        processReceiveTimeout();
+#endif
+
+    resumeFiber();
+}
+
+void ConnectionEstablisher::cancel()
+{
+    destroyFiber();
+    reset();
+}
+
+void ConnectionEstablisher::processReceiveTimeout()
+{
+#if defined(OS_LINUX)
+    destroyFiber();
+    stage = Stage::FAILED;
+    fail_message = "Code: 209, e.displayText() = DB::NetException: Timeout exceeded while reading from socket (" + result.entry->getDescription() + ")";
```

Comment:
Hm. Don't think we need `Code: 209, e.displayText() = DB::NetException` part here.

The first thing I want to do is to throw exception here. Probably, I cannot do it. Then there should be comment why.
BTW, may it be logical to distinguish timeout error from other error types?

Reply:
> Hm. Don't think we need Code: 209, e.displayText() = DB::NetException part here.

To be honest, I added this because of the integration test that checks this exception message)
Maybe I will just change the test.

Reply:
Ok :)
Yes, it is reasonable to change the test. E.g. check for substring `Timeout exceeded` only